pkg/aggregator/dist_sampler.go

```diff
@@ -8,83 +8,136 @@
-	return int64(timestamp) - int64(timestamp)%d.interval
+func (d *distSampler) truncateTs(ts int64) int64 {
+	return ts - ts%d.interval
 }
 
-// Add the metricSample to the correct sketch
-func (d *DistSampler) addSample(metricSample *metrics.MetricSample, timestamp float64) {
-	contextKey := d.contextResolver.trackContext(metricSample, timestamp)
-	bucketStart := d.calculateBucketStart(timestamp)
-	sketch, ok := d.sketchesByTimestamp[bucketStart]
-	if !ok {
-		sketch = metrics.MakeContextSketch()
-		d.sketchesByTimestamp[bucketStart] = sketch
+func (d *distSampler) addSample(ms *metrics.MetricSample, ts int64) {
+	ts = d.truncateTs(ts)
+	ck := d.ctxResolver.trackContext(ms, float64(ts))
+	d.m.insert(ts, ck, ms.Value)
+}
+
+func (d *distSampler) flush(flushTs int64) metrics.SketchSeriesList {
+	flushTs = d.truncateTs(flushTs)
+
+	pointsByCtx := make(map[ckey.ContextKey][]metrics.SketchPoint)
+	d.m.flushBefore(flushTs, func(ck ckey.ContextKey, p metrics.SketchPoint) {
+		if p.Sketch == nil {
+			return
+		}
+		pointsByCtx[ck] = append(pointsByCtx[ck], p)
+	})
+
+	out := make(metrics.SketchSeriesList, 0, len(pointsByCtx))
+	for ck, points := range pointsByCtx {
+		out = append(out, d.newSeries(ck, points))
 	}
-	sketch.AddSample(contextKey, metricSample, timestamp, d.interval)
+	d.ctxResolver.expireContexts(float64(flushTs) - defaultExpiry)
+	return out
 }
 
-func (d *DistSampler) flush(timestamp float64) percentile.SketchSeriesList {
-	var result []*percentile.SketchSeries
+func (d *distSampler) newSeries(ck ckey.ContextKey, points []metrics.SketchPoint) metrics.SketchSeries {
+	ctx := d.ctxResolver.contextsByKey[ck]
+	ss := metrics.SketchSeries{
+		Name:       ctx.Name,
+		Tags:       ctx.Tags,
+		Host:       ctx.Host,
+		Interval:   d.interval,
+		Points:     points,
+		ContextKey: ck,
+	}
+
+	if ss.Host == "" {
+		ss.Host = d.defaultHostname
+	}
+
+	return ss
+}
 
-	sketchesByContext := make(map[ckey.ContextKey]*percentile.SketchSeries)
+type sketchMap map[int64]map[ckey.ContextKey]*quantile.Agent
 
-	cutoffTime := d.calculateBucketStart(timestamp)
-	for bucketTimestamp, ctxSketch := range d.sketchesByTimestamp {
-		if cutoffTime <= bucketTimestamp {
+// Len returns the number of sketches stored
+func (m sketchMap) Len() int {
+	l := 0
+	for _, byCtx := range m {
+		l += len(byCtx)
+	}
+	return l
+}
+
+// insert v into a sketch for the given (ts, contextKey)
+// NOTE: ts is truncated to bucketSize
+func (m sketchMap) insert(ts int64, ck ckey.ContextKey, v float64) bool {
+	if math.IsInf(v, 0) || math.IsNaN(v) {
+		return false
+	}
+
+	m.getOrCreate(ts, ck).Insert(v)
+	return true
+}
+
+func (m sketchMap) getOrCreate(ts int64, ck ckey.ContextKey) *quantile.Agent {
+	// level 1: ts -> ctx
+	byCtx, ok := m[ts]
+	if !ok {
+		byCtx = make(map[ckey.ContextKey]*quantile.Agent)
+		m[ts] = byCtx
+	}
+
+	// level 2: ctx -> sketch
+	s, ok := byCtx[ck]
+	if !ok {
+		s = &quantile.Agent{}
+		byCtx[ck] = s
```

Comment:
TODO: maybe do `m[ts][ck] = s` here, not sure about nested maps and references if this resizes.